### PR TITLE
Add new catalog endpoint

### PR DESCRIPTION
**Pull request by pcarlisle**
very wip do not merge

### CODE REVIEW

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 1]
```diff
@@ -710,6 +712,48 @@
...
+  [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)]
+  (fn [request]
+    (let [{:keys [certname persistence trusted_facts facts environment classes parameters]}
+          (decode-catalog-post-body (slurp (:body request)))]
+      {:status 200
+       :headers {"Content-Type" "application/json"}
+       :body (json/encode
+              (jruby-protocol/compile-catalog jruby-service
+                                              (:jruby-instance request)
+                                              {:certname certname
```
Comment (tvpartytonight) [inline src/clj/puppetlabs/services/master/master_core.clj]: This map should probably contain all the keys listed in the schema in the `decode-catalog-post-body` method?

Comment (Magisus) [reply src/clj/puppetlabs/services/master/master_core.clj]: Agreed, I think they all end up being needed in the Ruby code.

File: src/ruby/puppetserver-lib/puppet/server/master.rb [hunk 2]
```diff
@@ -60,6 +60,10 @@ def handleRequest(request)
         response["X-Puppet-Version"])
   end
 
+  def compileCatalog(_request_data)
```
Comment (tvpartytonight) [inline src/ruby/puppetserver-lib/puppet/server/master.rb]: What is the reason for the `_` for `request_data`?

File: src/clj/puppetlabs/services/master/master_core.clj (hunk 1 above)
Comment (pcarlisle) [reply src/clj/puppetlabs/services/master/master_core.clj]: ok tony wrote that part

File: src/ruby/puppetserver-lib/puppet/server/master.rb (hunk 2 above)
Comment (pcarlisle) [reply src/ruby/puppetserver-lib/puppet/server/master.rb]: signifies that it's unused (for now)

**Comment by tvpartytonight** [comment]
I'll defer to @Magisus to actually merge this, since I wrote some of this too.

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 3]
```diff
@@ -710,6 +712,45 @@
...
+          (json/decode body true)
+          (catch JsonParseException e
+            (throw+ {:kind :bad-request
+                     :msg (format "Error parsing JSON: %s" e)})))]
+    (schema/validate
+     {:certname schema/Str
+      :persistence {:facts schema/Str, :catalogs schema/Str, :reports schema/Str}
+      (schema/optional-key :trusted_facts) {:values {schema/Any schema/Any}}
+      (schema/optional-key :facts) {:values {schema/Any schema/Any}}
+      (schema/optional-key :job-id) schema/Int
```
Comment (Magisus) [inline src/clj/puppetlabs/services/master/master_core.clj]: Can we choose either hyphen or underscores for these keys (this vs. `trusted_facts` above)? I don't have a strong preference which we choose.

Comment (pcarlisle) [reply src/clj/puppetlabs/services/master/master_core.clj]: @tvpartytonight where does this job-id come from?

Comment (pcarlisle) [reply src/clj/puppetlabs/services/master/master_core.clj]: looks like just an error, i've fixed it to _

**Comment by pcarlisle** [comment]
everything in here can probably be squashed

**Comment by Magisus** [comment]
Can maybe do that as part of the merge with the fancy button... Or you can squash things and then we merge.

**Comment by pcarlisle** [comment]
Feel free to use the button, either way is fine but that'll be faster if you're good with the changes. I just kept them separate for review purposes.

**Pull request merged**